### PR TITLE
Add Keras 3 example for "Transformer model for MIDI music generation"

### PR DESCRIPTION
**Pull request by johacks**
Hi,

This is my first contribution, so apologies in advance for mistakes made!

I saw a call for contribution with an example of midi generation using transformers. I adapted the [referenced code](https://github.com/jason9693/MusicTransformer-tensorflow2.0) to keras 3.

Here are some notes on the implementation:

- All MIDI datasets used by original code are inaccessible, so I switched to [Maestro dataset](https://magenta.tensorflow.org/datasets/maestro).
- I implemented the relative global attention used in the paper, which results in lots of additional code over using existing multi-head attention. Maybe it would be better to use already implemented attention layer `CachedMultiHeadAttention` to compact code, but probably getting worse results.
- Tokenization used in the code needs lots of code and is even hosted on a different repo. I have currently forked their repo and published to [pypi](https://pypi.org/project/midi-neural-processor/) so it can be easily installed, but I'm not sure this is the most appropiate way to handle it. Maybe it would make more sense to have a MIDI tokenizer on `keras_hub`?
- Following advice of Keras contributing guide [issue](https://github.com/keras-team/keras/issues/18442), I'm only including Python script before generating .md and .ipynb files.

Thanks!

### CODE REVIEW

File: examples/generative/midi_generation_with_transformer.py [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+
+import os
+import random
+import tempfile
+
+import keras
+import midi_neural_processor.processor as midi_tokenizer
+import numpy as np
+from keras import callbacks, layers, ops, optimizers, utils
+from keras_hub.api import layers as hub_layers
```
Comment (fchollet) [inline examples/generative/midi_generation_with_transformer.py:51]: Do `from keras_hub import layers as hub_layers`, don't import from `keras_hub.api`

File: examples/generative/midi_generation_with_transformer.py [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+### Hear a MIDI file
+
+We use the pretty_midi library and fluidsynth to convert MIDI files into waveform audio.
+This allows us to listen to the data samples before and after processing.
+"""
+
+
+def visualize_midi(midi_path, sampling_rate=16000, seconds=15, out_dir=None):
+    # Requires:
+    # sudo apt install -y fluidsynth
```
Comment (fchollet) [inline examples/generative/midi_generation_with_transformer.py:150]: Please add the installation instructions in the text rather than in a comment

File: examples/generative/midi_generation_with_transformer.py [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+The main difference is that it includes a relative positional encoding that
+allows the model to learn relative positional information between tokens.
+"""
+
+
+@keras.utils.register_keras_serializable()
+class RelativeGlobalAttention(layers.Layer):
+    """
+    From Music Transformer (Huang et al., 2018)
+    [paper link](https://arxiv.org/pdf/1809.04281.pdf)
```
Comment (fchollet) [inline examples/generative/midi_generation_with_transformer.py:316]: In a code example class docstring, links are never rendered so no need for markdown here

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. It looks great!

- Did you it works with all backends?
- Which backend is the fastest on the Colab GPU?
- Excluding text, how many lines of code do you end up with (are you able to run the rendering script)? You can use `scripts/tutobooks.py:count_locs_in_file` to count

**Comment by johacks** [comment]
@fchollet
Hi, thanks for the feedback, I have made corresponding changes according to comments on the code.

Regarding your other questions:
- _Did you check it works with all backends_?
  - I have now tried in in colab on all backends. I had to make some adjustments to the code:
      - In JAX it was too slow to concatenate multiple times for each generation, so I changed approach to fill a padded array with always the same size.
      - In torch there were some incompatibilities with negative slicing steps that have been fixed. It also seems that due to numerical stability issues, torch needs a lower learning rate. I adjusted the script to conditionally set a learning rate adjustment factor based on the backend.
-  _Which backend is the fastest on the Colab GPU_?
    - I tried all backends with a T4 GPU. It takes about 11 minutes for data preprocessing. Training + generation time is about 40 minutes, 27 minutes, **25 minutes** in Torch, JAX and **Tensorflow** respectively.
        - [Tensorflow notebook](https://colab.research.google.com/drive/1R5yMYo5L-msPidJ0LFp9_zLk_qny3RCF?usp=sharing)
        - [JAX notebook](https://colab.research.google.com/drive/1fF040k69tHM32w6JBz84vskVxcMTj9od?usp=sharing)
        - [Torch notebook](https://colab.research.google.com/drive/1CxZfIpj_ET-YjDDbaqDXQBpdyc9ortGC?usp=sharing)
- _Excluding text, how many lines of code do you end up with (are you able to run the rendering script)? You can use `scripts/tutobooks.py:count_locs_in_file` to count_
    - I was at 460. I refactored and got to 350.
    - I can generate the rendering script (`autogen.py add_example`), as long as it is launched with root privileges. I have added generated files to the PR.

**Review by fchollet — APPROVED** [review]
Awesome -- thanks a lot for the contribution!
